pkg/util/metrics.go

```diff
@@ -0,0 +1,279 @@
+package util
+
+import (
+	"fmt"
+	"github.com/golang/glog"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+	"k8s.io/client-go/util/workqueue"
+	"net/http"
+	"strings"
+	"sync"
+)
+
+type PrometheusMetrics struct {
+	Labels               []string
+	prefix               string
+	SparkAppSubmitCount  *prometheus.CounterVec
+	SparkAppSuccessCount *prometheus.CounterVec
+	SparkAppFailureCount *prometheus.CounterVec
+	SparkAppRunningCount *PositiveGauge
+
+	SparkAppSuccessExecutionTime *prometheus.SummaryVec
+	SparkAppFailureExecutionTime *prometheus.SummaryVec
+
+	SparkAppExecutorRunningCount *PositiveGauge
+	SparkAppExecutorFailureCount *prometheus.CounterVec
+	SparkAppExecutorSuccessCount *prometheus.CounterVec
+}
+
+func CreateValidMetric(prefix, name string) string {
+	// "-" aren't valid characters for prometheus metric names
+	return strings.Replace(prefix + name, "-", "_", -1)
+}
+
+type PositiveGauge struct {
```

Comment:
This is very interesting, but i think there are some potential gotchas. IF the cardinality of labels across a large time is high then you will have memory pressure as there is no purging. But, implementing purging is also not a good idea, because it will depend on the metrics server implementation.

I think we should add a comment. OVerall maintaining a guage may be very hard and it might be ok to have negative values